src/clients/wsrf/arcwsrf.1.in
.\" -*- nroff -*-
.TH ARCWSRF 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arcwsrf \- run WSRF query against service

.SH DESCRIPTION

The
.B arcwsrf
command issues WSRF resource property query to service specified by its URL.

.SH SYNOPSIS

.B arcwsrf [options] service_url [query]

.SH OPTIONS


  \-h, \-\-help                  Show help options
  \-p, \-\-property=[\-]name      Request for specific Resource Property
  \-t, \-\-timeout=seconds       timeout in seconds (default 20)
  \-z, \-\-conffile=filename     configuration file (default ~/.arc/client.conf)
  \-d, \-\-debug=debuglevel      FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
  \-v, \-\-version               print version information

.IP "\fB\-p\fR, \fB\-\-property\fR"
request specific Resource Property.
.IP "\fB\-t\fR, \fB\-\-timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB\-z\fR, \fB\-\-conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB\-d\fR, \fB\-\-debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB\-v\fR, \fB\-\-version\fR"
print version information
.IP "\fB\-?\fR, \fB\-\-help\fR"
print help

.SH ARGUMENTS

.IP "\fBservice_url\fR"
URL of service to which request is sent
.IP "\fBquery\fR"
query expression (usually XPath) to be sent
.LP

.SH EXTENDED DESCRIPTION

The
.B arcwsrf
send WSRF query to specified service URL and prints response. If query is
provided this query is sent to service. Otherwise request for whole
information document or for selected properties (if \-p options are specified)
is made.

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the
ARC client configuration file. By using the
.B \-\-conffile
option a different configuration file can be used than the default.

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT

APACHE LICENSE Version 2.0

.SH AUTHOR

ARC software is developed by the NorduGrid Collaboration 
(http://www.nordugrid.org), please consult the AUTHORS file distributed with 
ARC. Please report bugs and feature requests to http://bugzilla.nordugrid.org

.SH SEE ALSO

.BR arcinfo (1)


NEWS
no news good news :)


src/clients/wsrf/arcwsrf.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <list>
#include <string>

#include <arc/ArcLocation.h>
#include <arc/Utils.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/OptionParser.h>
#include <arc/StringConv.h>
#include <arc/UserConfig.h>
#include <arc/infosys/InformationInterface.h>
#include <arc/message/MCC.h>
#include <arc/communication/ClientInterface.h>

using namespace Arc;
#ifdef TEST
#define RUN(X) test_arcwsrf_##X
#else
#define RUN(X) X
#endif
int RUN(main)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::Logger logger(Arc::Logger::getRootLogger(), "arcwsrf");
  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  //Arc::ArcLocation::Init(argv[0]);
  Arc::OptionParser options(istring("URL [query]"),
                            istring("The arcwsrf command is used for "
                                    "obtaining the WS-ResourceProperties of\n"
                                    "services."));

  std::list<std::string> paths;
  options.AddOption('p', "property",
                    istring("Request for specific Resource Property"),
                    istring("[-]name"),
                    paths);

  int timeout = -1;
  options.AddOption('t', "timeout", istring("timeout in seconds (default 20)"),
                    istring("seconds"), timeout);

  std::string conffile;
  options.AddOption('z', "conffile",
                    istring("configuration file (default ~/.arc/client.conf)"),
                    istring("filename"), conffile);

  std::string debug;
  options.AddOption('d', "debug",
                    istring("FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG"),
                    istring("debuglevel"), debug);

  bool version = false;
  options.AddOption('v', "version", istring("print version information"),
                    version);

  std::string url;
  std::string query;
  {
    std::list<std::string> args = options.Parse(argc, argv);

    if (version) {
      std::cout << Arc::IString("%s version %s", "arcwsrf", VERSION)
                << std::endl;
      return 0;
    }

    if(args.size() < 1) {
      logger.msg(Arc::ERROR, "Missing URL");
      return 1;
    }

    if(url.size() > 2) {
      logger.msg(Arc::ERROR, "Too many parameters");
      return 1;
    }

    std::list<std::string>::iterator arg = args.begin();
    url = *arg;
    ++arg;
    if(arg != args.end()) query = *arg;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::istring_to_level(debug));

  Arc::UserConfig usercfg(conffile);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (timeout > 0)
    usercfg.Timeout(timeout);

  if (debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::istring_to_level(usercfg.Verbosity()));

  //if (timeout > 0) {
  //  usercfg.SetTimeout(timeout);
  //}

  // Proxy check
  //if (!usercfg.CheckProxy())
  //  return 1;

  CountedPointer<InformationRequest> request(NULL);
  if(!query.empty()) {
    XMLNode q(query);
    if(!q) {
      logger.msg(Arc::ERROR, "Query is not a valid XML");
      return 1;
    }
    request = new InformationRequest(q);
  } else {
    std::list<std::list<std::string> > qpaths;
    for(std::list<std::string>::iterator path = paths.begin();
                                         path != paths.end();++path) {
      std::list<std::string> qpath;
      qpath.push_back(*path);
      qpaths.push_back(qpath);
    }
    request = new InformationRequest(qpaths);
  }
  if(!(*request)) {
    logger.msg(Arc::ERROR, "Failed to create WSRP request");
    return 1;
  }
  //SOAPEnvelope* SOAP(void);

  URL u(url);
  if(!u) {
    logger.msg(Arc::ERROR, "Specified URL is not valid");
    return 1;
  }

  MCCConfig cfg;
  usercfg.ApplyToConfig(cfg);
  ClientSOAP client(cfg, u, usercfg.Timeout());
  PayloadSOAP req(*(request->SOAP()));
  PayloadSOAP* resp_ = NULL;
  MCC_Status r = client.process(&req,&resp_);
  CountedPointer<PayloadSOAP> resp(resp_);
  if(!r) {
    logger.msg(Arc::ERROR, "Failed to send request");
    return 1;
  }
  if(!resp) {
    logger.msg(Arc::ERROR, "Failed to obtain SOAP response");
    return 1;
  }
  std::string o;
  resp->Body().Child().GetXML(o);
  if(resp->IsFault()) {
    logger.msg(Arc::ERROR, "SOAP fault received");
    std::cerr<<o<<std::endl;
    return 1;
  }
  std::cout<<o<<std::endl;
  return 0;
}
